### PR TITLE
Edit Site: Add convert to template part flow.

### PR DESCRIPTION
**Pull request by epiqueras**
## Description

This PR makes it possible to select and convert blocks to a template part in the site editor, similar to how it works for the group and reusable blocks.

It also fixes the breakage that G2 introduced to the block toolbar within the site editor.

## How has this been tested?

It was verified that selecting blocks in the site editor, clicking on the block toolbar's settings dropdown, and selecting the option to make a template part, creates a template part with the chosen blocks after inserting the new template part's name and theme.

## Screenshots

gif

## Types of Changes

*New Feature:* The site editor now has a button in the block toolbar's settings dropdown for converting a selection of blocks into a template part.

## Checklist:

- [ ] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by mtias** [comment]
This is great. It starts to tie in experiences that were laid back with reusable blocks. I expect this to be a very common way to build a theme or template — start by designing with blocks, then group the pieces that make sense for different template parts.

File: packages/block-library/src/template-part/edit/inner-blocks.js [hunk 1]
```diff
@@ -14,7 +14,8 @@ export default function TemplatePartInnerBlocks() {
 	);
 	return (
 		<InnerBlocks
-			__experimentalBlocks={ blocks }
+			// Allow initial blocks from a block transform to load, if any.
+			__experimentalBlocks={ blocks.length === 0 ? undefined : blocks }
```
Comment (noahtallen) [inline packages/block-library/src/template-part/edit/inner-blocks.js:18]: this should be `value={ }` now :) I'm also not sure if it's still needed with the latest block sync changes

File: packages/blocks/README.md [hunk 2]
```diff
@@ -730,6 +730,7 @@ _Parameters_
 
 -   _blocks_ `(Array|Object)`: Blocks array or block object.
 -   _name_ `string`: Block name.
+-   _isGroupingBlock_ `boolean`: Set to true to bypass default grouping block selection.
```
Comment (noahtallen) [inline packages/blocks/README.md:733]: What does this do? Is bock transform normally wrapped by a group block or something?

File: packages/edit-site/src/components/template-part-converter/index.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
+/**
+ * WordPress dependencies
+ */
+import { useSelect, useDispatch } from '@wordpress/data';
+import { BlockSettingsMenuControls } from '@wordpress/block-editor';
+import { MenuItem } from '@wordpress/components';
+import { switchToBlockType } from '@wordpress/blocks';
+import { __ } from '@wordpress/i18n';
+
+export default function TemplatePartConverter() {
```
Comment (noahtallen) [inline packages/edit-site/src/components/template-part-converter/index.js:10]: Should this live in the `block-editor` package? It seems like it would fit well there, since it's basically part of the block menu. (maybe we could then provide a mechanism to turn it off for the post editor?)

**Review by noahtallen — COMMENTED** [review]
just some thoughts and questions. It also needs rebased and tested locally again, but I'm sure you already know that ;)

**Comment by MichaelArestad** [comment]
I dig this. We can change the text from "template part" to whatever we decide to call them. I know I'm leaning toward "Section".

**Comment by mtias** [comment]
@MichaelArestad would we still call it "section" if it's something like just a navigation menu, or a list of social icons?

**Comment by MichaelArestad** [comment]
> @MichaelArestad would we still call it "section" if it's something like just a navigation menu, or a list of social icons?

Good question. I think section still works in those scenarios.

**Comment by Addison-Stavlo** [comment]
This convert-to flow seems like it will be super-smooth when combined with some of the other creation changes being developed in #23295!

Once these flows combine, will we be able to 'undo' the conversion if necessary (and also deleting the auto-draft created for the 'Untitled Section' in that undo) ?

<hr/>

@matias mentioned:
> would we still call it "section" if it's something like just a navigation menu, or a list of social icons?

It sounds like we have pretty much settled on the term 'Section' ?  If thats what we have decided to move forward with then I don't expect to change anyones mind, but I do have some fair concerns regarding this name:

1.  'Section' and 'Group' are synonymous.   So much so that searching for 'Section' in the block inserter brings the 'Group' block up as the first result.  If there is already a subset of users that think of the 'Group' block as a section, this will be overly confusing.  By the names and previous convention, a user would expect the Section block and Group block to behave very similarly.

2. When we get around to adding semantic tag support for template parts, `<section>` would be one of those options.  If we are calling this block a "Section", then the Section block would have the option to have a 'section' tag (as well as 'header', 'footer', etc.), creating a false implication that the 'section' tag would be the preferred tag for the Section block.  This seems like an unfortunate confusion and redundancy. 

3. 'Section' conveys zero meaning towards the behavioral properties and reusability of the saved entity.  While the term 'Template Part' may be on the 'too technical' side of the spectrum, 'Section' seems to be on the exact opposite, 'too vague' side.  If a user comes across a 'too technical' term that they do not understand, they are generally prompted to search for its meaning.  However, when a user comes across a 'too vague' term, they could often assume they understand the meaning and consequently mis-use the block and get themselves in a bit of a mess.  For this reason, 'too technical' seems preferable to 'too vague'.

4.  A complete disconnection between terms used in theme development and site editing.  Calling it 'Thing A' in one place and 'Stuff B' in another is asking for more confusion points.

For these reasons I feel that renaming this 'Section' in the editor will cause a handful of different confusion points down the road and be looked back upon as an unfortunate naming convention.

**Comment by MichaelArestad** [comment]
> For these reasons I feel that renaming this 'Section' in the editor will cause a handful of different confusion points down the road and be looked back upon as an unfortunate naming convention.

These are all good points. What do you propose?

**Comment by Addison-Stavlo** [comment]
> These are all good points. What do you propose?

TLDR: To not rename it.

Consider a standard user who we are concerned with exposing the term 'Template Part' to:
* Going into the site editor will load a Template from their theme.
* That Template will already have 'Template Parts' in place.
* If the user doesn't like these Template Parts, they can click on them and edit them.  Or use an interface from the toolbar to select a different one.
* At no point in this process has the user ever had to understand, know, or really be exposed to the term 'Template Part'.

If a user needs to find the "Template Part" block, it will be to add it to a Template.  In this case the term 'Template Part' makes sense.  If a user is editing Templates in this manner they are now leaning towards the side of theme development, in which case they should be exposed to the proper term that Themes use.  (This assumes we wouldn't use the template part block in the post editor, but I believe this was suggested by both yourself and @mtias in the past? As noted we do have "Reusable Blocks" for that).

I personally don't think 'Template Part' really needs to be renamed.  It seems like it would cause more [confusion](https://github.com/WordPress/gutenberg/pull/20445#issuecomment-649754693) than clarity.  I really have tried to think of a "good alternative" for the sake of the discussions but I don't feel that there really is any good alternative where the pros outweigh the cons.  If it really needs to be renamed, something other than 'Section' would cut out a couple of the confusion points (but not all of them).

**Comment by MichaelArestad** [comment]
@Addison-Stavlo I'm sold on going back to Template Parts. You make excellent points with benefits that outweigh the reasons I wanted to rename it in the first place. Should I create an issue or would you be interested in changing it back?

**Comment by Addison-Stavlo** [comment]
> Should I create an issue or would you be interested in changing it back?

Go ahead and create an Issue, I'm stuck on other priorities/duties until the end of next week.  It should be a fairly quick task, but if its still around by then I can get to it.

File: packages/edit-site/src/components/block-editor/index.js [hunk 4]
```diff
@@ -49,6 +50,7 @@ export default function BlockEditor( { setIsInserterOpen } ) {
 			useSubRegistry={ false }
 		>
 			<BlockEditorKeyboardShortcuts />
+			<TemplatePartConverter />
```
Comment (Addison-Stavlo) [inline packages/edit-site/src/components/block-editor/index.js:53]: This component and its corresponding file seem unnecessary if we are fine converting from the standard block transform flow:
Screen Shot 2020-10-20 at 5 08 45 PM

File: packages/block-library/src/template-part/edit/placeholder/index.js [hunk 5]
```diff
@@ -35,6 +40,17 @@ export default function TemplatePartPlaceholder( { setAttributes } ) {
...
+	// If there are inner blocks already present, this is a convert to template part flow.
+	// Therefore immediately create the template part with the given inner blocks.
+	useEffect( () => {
+		if ( innerBlocks.length ) {
+			onCreate();
+		}
+	}, [] );
+	if ( innerBlocks.length ) {
+		return <Spinner />;
+	}
```
Comment (Addison-Stavlo) [inline packages/block-library/src/template-part/edit/placeholder/index.js:52]: Since the placeholder has updated a lot since this PR was originally made, I had to make some changes for this transform to work properly.  Here, we pass down the innerBlocks to the placeholder and immediately create the entity if they exist.  The idea is that this block will never render the placeholder when innerBlocks are present, unless we are in the conversion flow.

**Review by Addison-Stavlo — COMMENTED** [review]
This PR should be in a working state again:

convert-to-template-part


I am noticing this PR adds the ability to convert to template part in 2 separate areas of the toolbar, the new button as well as the standard block-transform tool.

In transform menu:
Screen Shot 2020-10-20 at 5 08 45 PM

New button:
Screen Shot 2020-10-20 at 5 15 52 PM

@shaunandrews  @jameskoster  @MichaelArestad - Are we fine with keeping it in the transform tool and getting rid of the separate button/component?  Or do we want to somehow hide it from the transform tool and only allow it to happen from this new button area?  (The transform tool seems to appear auto-magically once the transform is defined, which the separate button still relies on for functionality.  So I am not certain the best way to hide it if we don't approve of it in the standard transform location). 🤔

**Comment by jameskoster** [comment]
I think it's fine in the transform tool. If grouping is not available via the ellipsis menu, I probably wouldn't place the template part tool there either.

**Comment by mtias** [comment]
Grouping is different. The equivalent here is "Convert to reusable block", which is only available in the ellipsis menu.

**Comment by Addison-Stavlo** [comment]
> The equivalent here is "Convert to reusable block", which is only available in the ellipsis menu.

Updated to have this only available in the ellipsis menu.  Should this be a site-editor only feature?  Or should it be available in the post editor as well with the FSE experiment enabled?

File: packages/edit-site/src/components/block-editor/index.js (hunk 4 above)
Comment (Addison-Stavlo) [reply packages/edit-site/src/components/block-editor/index.js:53]: We will need this as we got rid of the transform flow since this should only be available through the ellipsis menu - https://github.com/WordPress/gutenberg/pull/20445#issuecomment-713499116

File: packages/edit-site/src/components/template-part-converter/index.js (hunk 3 above)
Comment (Addison-Stavlo) [reply packages/edit-site/src/components/template-part-converter/index.js:10]: I would assume that if we want this in both editors (site and post) when FSE is enabled, that this would be better in `block-editor` as well.  However, if it doesn't belong in the post editor I would think this is a fitting location for it?  (asked about this below)

**Comment by Addison-Stavlo** [comment]
This functionality is staying edit-site only for now.  So this should be g2g for review! 😁

File: packages/edit-site/src/components/template-part-converter/index.js (hunk 3 above)
Comment (Addison-Stavlo) [reply packages/edit-site/src/components/template-part-converter/index.js:10]: It looks like the idea is to keep it edit-site only (for now at least).  Id say edit-site package makes sense, and we can move it over to block-editor if things change and we want this in the post editor as well.

**Comment by jeyip** [comment]
### Requirements
- "Make Template Part" button in the transform menu is only available in the block toolbar ellipses popover
- Can create template part from a single block
- Can create template part from multiple blocks
- Can create template part from multiple template parts (had to use select mode to focus multiple template parts)

### Browsers
- [ ] Chrome
- [ ] Edge
- [ ] Firefox
- [ ] Safari
- [ ] IE11 (Having trouble getting local environment running)

**Review by jeyip — COMMENTED** [review]
The functionality works well 👍 . It does, however, feel unintuitive to me that the logic for the convert to template part flow lives in the template-part placeholder. It also seems odd that the Placeholder needs to know about `innerBlocks`. Giving this more thought at the moment.

Would it be possible to save the new template part entity record [when we are clicking on "Make template part" button](https://github.com/WordPress/gutenberg/pull/20445/files#diff-96f4a7c0b163432cb5b21d2bfabdd0aa2f69ad6ef75b348b1ff577fd6b16cc8aR32-R48), instead of trying to reuse the logic in the placeholder? 

Another alternative might be to have the `onCreate` callback and convert to template part flow logic live in `template-part/edit`. `onCreate` could be prop drilled into the placeholder. 

I'm currently thinking about if any of this is possible, what they would entail, and if they're worthwhile to implement. Curious about your thoughts @Addison-Stavlo

**Comment by Addison-Stavlo** [comment]
> Would it be possible to save the new template part entity record when we are clicking on "Make template part" button, instead of trying to reuse the logic in the placeholder?

Yeah, you are right this would probably make more sense.

**Comment by Addison-Stavlo** [comment]
> Yeah, you are right this would probably make more sense.

🤔   Or does it?  If we replace the blocks with a template part block immediately then let the template part block take care of creation, then we have a way for that block to provide feedback (like a loading spinner) while that creation is in progress.  If we create it at the 'make template part' button, then the block switch would await that entity creation causing a delay where it isn't very apparent something is happening?  Maybe there is a way to get around this.

But if we are sending the innerBlocks off to the template part block to handle creation, then the 'placeholder' sort of makes sense to handle triggering that creation, since it is a placeholder for when the entity does not yet exist.

**Comment by jeyip** [comment]
> If we create it at the 'make template part' button, then the block switch would await that entity creation causing a delay where it isn't very apparent something is happening? Maybe there is a way to get around this. Maybe there is a way to get around this.

Oh that's an interesting point. I wonder if there is a generic loading state for blocks that are in the process of being created. With Matias's comments earlier, it sounds like the closest parallel to this flow is when we create reusable blocks. I'll take a look at that.

> But if we are sending the innerBlocks off to the template part block to handle creation then the 'placeholder' sort of makes sense to handle triggering that creation, since it is a placeholder for when the entity does not yet exist.

If we go down this route, then it would be nice to move the onCreate callback and template part flow logic up to its parent. The make template part flow would be more visible and it would remove the innerBlocks dependency from the placeholder. To be clear though, I haven't tried this yet, and I'm not sure if it would be possible. Also curious about your opinions on this Addie.

**Comment by jeyip** [comment]
For the "Add to Reusable Blocks" flow in post editing, we use a spinner and a notice to provide user feedback. I imagine that they're handling spinner rendering through redux ui state, but I'll have to dig more.

`packages/reusable-blocks/src/components/reusable-blocks-menu-items/reusable-block-convert-button.js:77`

**Comment by jeyip** [comment]
Yeah it looks like it uses resolution selectors from `wordpress/data` to render the spinner

`packages/block-library/src/block/edit.js:47`
`packages/block-library/src/block/edit.js:100`

**Comment by Addison-Stavlo** [comment]
> I imagine that they're handling spinner rendering through redux ui state, and I think we could probably follow the same pattern for template parts

It is similar.  The spinner is rendered by the block as an early return if the entity is not yet resolved.  Reusable blocks has its entire separate package with its own data store as well, its a bit to get into.  We could potentially do something like this for template parts in the future if it is necessary, but I was hoping to avoid that for now.

> If we go down this route, then it would be nice to move the onCreate callback and template part flow logic up to its parent. The make template part flow would be more visible 

Im having trouble understanding the benefits of this. 🤔   Why does the parent need that logic?  The placeholder is what is rendered when there is no template part entity corresponding to the block's attributes, and as a result it is where the creation/selection flow stems from.  Why would the onCreate function then live higher up the chain?  Im not sure what you mean by it being 'more visible'.

> and it would remove the innerBlocks dependency from the placeholder. 

What does this do for us?  The placeholder isn't dependent on innerBlocks, it either receives them or it doesn't.  If it does, the context of creation is already clear as the content for the new template part is already selected and defined.

**Comment by Addison-Stavlo** [comment]
> Yeah it looks like we use resolution selectors from the wordpress/data redux store to render the spinner

In reusable blocks case, I don't think there is a 'placeholder' block to select or create a blank reusable block.  Blocks are converted to a reusable block, and they are displayed as individual options to insert from the block inserter.  When the controlling block mounts it has a record that it corresponds to, which means it can check that record and display a spinner until it is available

However, with the template part our placeholder has options to select or create blank.  When the controlling block mounts, there may or may not be a corresponding record to expect, and thats why the options to select/create exist from that placeholder.  So if we get to the placeholder and check redux for a resolved record, we would also need some extra context to know whether we should display a spinner if it is unresolved, or to display the standard select/create blank if we shouldn't be expecting a record to resolve in the first place? 🤔   The innerBlocks passed to the placeholder is a clear indication of that use context and makes that determination simple.

**Comment by jeyip** [comment]
Hmm...there are a few threads of conversation here. To quickly summarize and make sure we're on the same page:

All of these discussions originate from the fact that, to me, it felt odd to have the convert template part flow logic inside of the template part placeholder.


**Discussion 1**
[We think the placeholder may in fact be the right place to trigger creation template part creation](https://github.com/WordPress/gutenberg/pull/20445#issuecomment-714802089) and the PR should be left as is.

**Discussion 2**
Template part creation could happen immediately when we click on the template part converter button, but it will cause a delay (without user feedback) as we await entity creation. Maybe we could follow the pattern established by reusable blocks?
  - We could address the delay by placing ui state in redux so that a loading state can be rendered immediately, as is done with reusable blocks, although [it might involve a lot of work ](https://github.com/WordPress/gutenberg/pull/20445#issuecomment-714821967)
  - [Reusable blocks also behave differently than template parts](https://github.com/WordPress/gutenberg/pull/20445#issuecomment-714828260)

**Discussion 3**
We're talking about moving the convert template part flow business logic and the placeholder onCreate callback into its parent.
  - What value does this bring?

**Comment by jeyip** [comment]
**We think the placeholder may in fact be the right place to trigger creation template part creation and the PR should be left as is.**



**Template part creation could happen immediately when we click on the template part converter button, but it will cause a delay (without user feedback) as we await entity creation.**

Apologies if this discussion was misleading. When I looked at reusable blocks, I didn't mean to say that we should create an entirely new store (as is done with reusable blocks). 

I was trying to convey the idea that, if we save a new template part entity record immediately when clicking the "Make template part" button, I thought we could track loading state with a resolution selector like `hasFinishedResolution` or `isResolving` selector, allowing us to provide user feedback during the delay.

You mentioned earlier that the "block switch would have to await entity creation", but I might be missing something. I'm not sure why the block switch has to wait for entity creation. Doesn't [replaceBlocks](https://github.com/WordPress/gutenberg/pull/20445/files#diff-96f4a7c0b163432cb5b21d2bfabdd0aa2f69ad6ef75b348b1ff577fd6b16cc8aR33) create a template part block, at which point we could use a resolution selector to display a loading spinner? (Again might be misunderstanding something here)

**We're talking about moving the convert template part flow business logic and the placeholder onCreate callback into its parent.**

**Comment by jeyip** [comment]
**We're talking about moving the convert template part flow business logic and the placeholder onCreate callback into its parent.**

You asked a number of great questions in this discussion. To frame my reasoning, I'll start by describing why I thought it could be beneficial to move `innerBlocks` to the parent.

> What does this do for us? The placeholder isn't dependent on innerBlocks, it either receives them or it doesn't. If it does, the context of creation is already clear as the content for the new template part is already selected and defined.

I think we have two different definitions of dependency here. I was trying to say that `innerBlocks` is an argument, or prop, that is passed to the Template Part Placeholder. It's necessary for the Placeholder to behave properly (knowing when to call onCreate, render a spinner, etc.). 

I was thinking that if we moved `innerBlocks` into its parent, we could also, as a result, separate concerns, keep the responsibilities of the placeholder narrowly defined, and move all business logic into the parent. The Placeholder w

We already see this emerging as a pattern, which is to say that `template-part/edit` because it contains the business logic of when to render the placeholder vs. the name panel + dropdown vs. inner blocks vs. a spinner.

**Comment by Addison-Stavlo** [comment]
Re: Creating the entity in the converter component vs. template part block:

>  if we save a new template part entity record immediately when clicking the "Make template part" button, I thought we could track loading state with a resolution selector like hasFinishedResolution or isResolving selector, allowing us to provide user feedback during the delay.

Right, we could use state to track this information, however it would require extra complexity and reworking how the template part block works to enable this in a meaningful way that still allows the standard select/create flows to work and appear when expected vs. the 'loading spinner' that is used to represent waiting a creation that was triggered elsewhere.  

While reusable blocks is our closest equivalent to model this flow after, there are enough differences in the overall functionalities that we shouldn't expect the underlying code for this flow to look the same.  The block controlling reusable blocks has no concept of creation/selection, while the template part block does.  Since it does, I think it makes much more sense to keep using this functionality to create the entity after block conversion than to define a new separate creation system elsewhere and work to add extra logic to determine whether the template part block was initiated via a conversion awaiting entity creation elsewhere vs. being added from the inserter.  The block shouldn't care about how it was put on the page, it should just initialize and make its determinations in how to function without needing that extra information.

The state of this PR back in February did this sort of flow as well.  It converted block(s) to innerBlocks of a template part block that represented no entity, the template part block mounted and found no corresponding saved entity and therefore loaded the placeholder, the placeholder was then used to give the slug/theme names in an input and click 'create', and creation was done through the placeholder using the innerBlocks that were present.  I think this still makes sense today, the only difference here is that the placeholder no longer asks for slug/theme inputs in creating a new template part as it was updated to create all new custom template parts as 'Untitled Template Part'.  Since that creation flow was simplified, this PR has been updated to trigger that same creation from that same place when it is necessary.

<hr/>

Re: Moving onCreate/innerBlocks out of the placeholder:

> I was thinking that if we moved innerBlocks into its parent, we could also, as a result, separate concerns,

I feel the concerns are separated in this way.  When the template part block cannot find an entity, the placeholder is rendered and handles what needs to happen for the creation of that entity.  Moving some of that creation to one place and leaving the rest of it in another seems like over-separation.  That is, we are splitting the same concern up into multiple places instead of having it consolidated into one.

> We already see a lot of business logic being consolidated in template-part/edit. For example, it already decides when to render the placeholder vs. the name panel + dropdown vs. inner blocks vs. a spinner.

Right, basically I see the current state of that edit file as:
* If no corresponding entity -> render placeholder that handles creation
* If an entity is resolved -> toolbar controls + inner blocks
* if unresolvable file -> some error spinner

Agreed, there is a lot of business logic there.  But I don't see that as a reason to clutter it with more logic if it's not necessary. If the template part block mounts with a corresponding entity it has no need of having an `onCreate` function defined, so to me it makes sense to not define this until we get to the component that handles creation.  If we wanted to separate the creation with existing inner blocks from the select/create flow in the placeholder, that would become something like:

* first move all creation logic here instead of the placeholder
* If no corresponding entity + innerBlocks present -> trigger a creation
* If no corresponding entity -> render placeholder that also handles creation in other cases
* If an entity is resolved -> toolbar controls + inner blocks
* if unresolvable file -> some error spinner

We could do that, but to me it seems more cluttered and less clearly defined.

**Comment by jeyip** [comment]
> it would require extra complexity and reworking how the template part block works to enable this in a meaningful way that still allows the standard select/create flows to work and appear when expected vs. the 'loading spinner' that is used to represent waiting a creation that was triggered elsewhere.

Gotcha 👍 That's the sense I got from previous comments, but I wanted to confirm to make sure we're that I wasn't misunderstanding. Thanks!

**Comment by jeyip** [comment]
> The block controlling reusable blocks has no concept of creation/selection, while the template part block does. Since it does, I think it makes much more sense to keep using this functionality to create the entity after block conversion than to define a new separate creation system elsewhere and work to add extra logic to determine whether the template part block was initiated via a conversion awaiting entity creation elsewhere vs. being added from the inserter. The block shouldn't care about how it was put on the page, it should just initialize and make its determinations in how to function without needing that extra information.

The point about a block not caring about how it was put on a page is very interesting. I didn't realize that's a typical expectation. I think I'm tracking. Correct me if I'm wrong, but you're saying that the block should act as its own self-contained unit, and when we say make its determinations of how to function, that includes triggering template part creation.

**Comment by Addison-Stavlo** [comment]
>  I didn't realize that's a typical expectation. I think I'm tracking. Correct me if I'm wrong, but you're saying that the block should act as its own self-contained unit, and when we say make its determinations of how to function, that includes triggering template part creation.

I think it is preferable, if the block is set up to handle these determinations I don't see a reason to complicate it by having it rely on more information about how it was added to the editor.

**Comment by jeyip** [comment]
> The state of this PR back in February did this sort of flow as well.

> I think this still makes sense today, the only difference here is that the placeholder no longer asks for slug/theme inputs in creating a new template part as it was updated to create all new custom template parts as 'Untitled Template Part'. 

I see. Enrique created this a while ago, had a certain intention when structuring the PR the way he did, was also handling the convert flow in the placeholder, and was, in fact, requiring even more information than it is now (slug + theme inputs).

**Comment by Addison-Stavlo** [comment]
Yeah, thats how I see it.

> was also handling the convert flow in the placeholder

To be clear, the 'convert flow' is and has been happening outside of the block.  It is what replaces the blocks with the template part block in the first place.  By that point the blocks are technically converted, but the template part block and its corresponding inner blocks that now exist don't correspond to any entity, and that entity creation is still handled by the placeholder.

**Comment by Addison-Stavlo** [comment]
I think the creation from innerBlocks in the placeholder adds a bit of a protection as well in some cases.  Perhaps there is a template part block with inner blocks in the content passed into the editor that does not have a corresponding entity (maybe it was defined in markup in an unconventional way with no actual corresponding file, or maybe some other circumstance).  In that case, instead of the block failing and either loading the placeholder or the final unresolved file spinner, it would automatically trigger creation of 'Untitled Template Part' and the content would still be available as expected. 🤔

**Comment by jeyip** [comment]
> When the template part block cannot find an entity, the placeholder is rendered and handles what needs to happen for the creation of that entity. Moving some of that creation to one place and leaving the rest of it in another seems like over-separation.

> Agreed, there is a lot of business logic there. But I don't see that as a reason to clutter it with more logic if it's not necessary. 

I think I see what you're conveying here. Correct me if I'm wrong, but we're saying that the concerns that we're separating are:
  
1. the inability for a template part block to find an entity
2. template part creation itself.

This is where I disagree, but at the same time, I want to acknowledge that I might frame the concerns in separation of concerns differently because of past experiences at other companies (which may not translate to Automattic). To me it feels more natural to consolidate all business logic in the parent, but the more I think about it, the more I see my disagreements may also be a matter of opinion.

Ultimately, the functionality in this PR works well, I believe I understand your reasoning more clearly (chime in if you feel like I'm missing anything big! 🙏 ), and all of my questions have been resolved, so +1 from me 🚀

**Review by jeyip — APPROVED** [review]

**Comment by jeyip** [comment]
> Just another thought - I think the creation from innerBlocks could add a bit of a protection/fallback as well in some cases.

Oh also an interesting point! Giving this some thought 🤔

**Comment by Addison-Stavlo** [comment]
> Correct me if I'm wrong, but we're saying that the concerns that we're separating are:
> * the inability for a template part block to find an entity
> * template part creation itself.

Thats the way I see how the concerns should be separated:
* Parent -> concerned with finding the entity, and whether to display the entity vs. the placeholder
* Placeholder -> concerned with determining when/how to create or select the entity

I think what you are proposing would be more:
* Parent -> concerned with finding the entity, and determining what type of creation may need to happen
* Parent (or new child component) -> concerned with triggering creation involving pre-existing inner content.
* Placeholder -> concerned only with manual selection/blank entity creation.

I think both are valid and will work, and we have a different opinion on which makes more sense?

**Comment by jeyip** [comment]
> I think both are valid and will work, and we have a different opinion on which makes more sense?

Yeah! I think that's a good way to describe this. I agree that both ways will work 👍 It's a difference of opinion, and I definitely don't think that should stop this PR from being merged.

**Comment by mtias** [comment]
Thanks for pushing this one through!
